### ISSUE TITLE
 Subroutines::RequireFinalReturn should allow a method call in terminal_funcs

### ISSUE THREAD

**Issue by robinsmidsrod** [Perl-Critic/Perl-Critic#920]
If you throw an exception using a method call, like `MyClass->throw_exception(...)` the `_is_terminal_stmnt` sub only considers the first token, and not the entire method call. That means that we would have to write `terminal_funcs = MyClass` for this to be ignored, and that would not be a specific enough exemption.

It would be very beneficial if it was possible to also define a class and/or instance method that throw exceptions.

**Comment by xenomorpheus** [comment]
TL;DR RequireFinalReturn isn't able to recognise **methods** that are **terminal**, like it is able to recognise subroutines that are terminal.

Details as per submission process:

Pull request created: https://github.com/Perl-Critic/Perl-Critic/pull/935

1. How can it be reproduced

In a subroutine use a terminal statement such as logconfess from Log4Perl.
Please see attached [Example.pm.txt](https://github.com/Perl-Critic/Perl-Critic/files/5719386/Example.pm.txt)

```
   $LOGGER->logconfess("just a test");
} ## End of my_sub
```

perlcritic --profile .perlcriticrc -severity 3 Example.pm 
Subroutine "my_sub" does not end with "return" at line 8, column 1.  See page 197 of PBP.  (Severity: 4)

2. What was expected

Expect that **logconfess** method be treated as a terminal statement, just like croak, confess, die, ...

3. What actually occurred

logconfess method wasn't recognised as terminal.

4. What version of the involved component

1.139_01 in RequireFinalReturn.pm

Additional:
Follows existing code pattern. Additional method names may be suppled as per following exmaple.

```
[Subroutines::RequireFinalReturn]
terminal_meths = logconfess
```

**Comment by petdance** [comment]
The argument should be `terminal_methods`, not `terminal_meths`.

Any PR needs to have tests to verify it works, both positive and negative.

We need to have updated docs to tell users about the feature.  

I'm concerned about declaring a terminal method that conflicts with another, non-terminal method in another class. What are your thoughts on that?

**Comment by robinsmidsrod** [comment]
@petdance Going back to my initial example, I would expect to define `terminal_methods = MyClass::throw_exception` in my .perlcriticrc (or `->`, if `::` would be somewhat confusing).

That would mean that we probably only would be able to detect class methods invoked like this `MyClass->throw_exception(...)` and not instance invocation, like `$my_instance->throw_exception(...)`.

Is that an acceptable limitation? If you don't specify any class, then the method name would match for any class.

Another option could be to write something like this `terminal_methods = $my_instance->throw_exception` to allow that exact variable name for the instance to not generate a policy violation. That way, both a class and instance invocation could be supported, if the thing before the arrow is just literally matched against the source code element. After all, this is static analysis...

**Comment by petdance** [comment]
I don't think tying it to a specific variable name like `$my_instance->throw_exception` is the way to go.  There's nowhere else we do anything like that.

The `MyClass::throw_exception` could be possible, wouldn't solve your original problem of `$LOGGER->logconfess("just a test");`, and wouldn't be able to work with any derived classes.

It seems to me that there's really not any way to do this other than saying "If you define `throw_exception` as terminal, then all of them will be, regardless of class", and make that explicitly clear in the docs (another reason docs are crucial on this).

**Comment by robinsmidsrod** [comment]
@petdance I'm okay with the interpretation of not caring what the associated class is. As you say, it's the best we can do at this stage. As long as we clearly document how it behaves, that should be good enough, right?

@xenomorpheus Any additional insight to share on your use-case?

**Comment by petdance** [comment]
> As long as we clearly document how it behaves, that should be good enough, right?

Agreed.  I'd be OK with this getting added once we get docs and tests.

**Comment by xenomorpheus** [comment]
> The argument should be `terminal_methods`, not `terminal_meths`.
The established style of this package uses `terminal_funcs`, so rather than have a mix of styles (adding to confusion) naming  followed suit and used `terminal_meths`.  Typically good software favours consistency / conformity.
If you aren't swayed by this reasoned argument I will of course change to the proposed `terminal_methods`.  Please let me know your decision.

It is very late in my timezone, so I'll try understand and implement the docs and tests over the next few days.

**Comment by xenomorpheus** [comment]
@petdance @robinsmidsrod who needs sleep :) I'll try to respond to more of the comments.

> @petdance I'm okay with the interpretation of not caring what the associated class is. As you say, it's the best we can do at this stage. As long as we clearly document how it behaves, that should be good enough, right?

@petdance @robinsmidsrod Agreed. I think (when supported by doco) this is the best we can do as the test is a static analysis, not a run-time.  If I understand you correctly, I feel what we proposed is what I have implemented. Please let me know if I have misunderstood something.

Suggestions on doco very much welcome. To work from your text.

"If you define throw_exception as terminal, then any method of that name will be terminal, regardless of class."

I'll have to go at adding that to the existing doco.

> @xenomorpheus Any additional insight to share on your use-case?

@petdance I'll have a guess at answering your question. Please let me know if I have misunderstood you.

In the example code I have `my_sub`, a subroutine that fails to end in a return statement.  It also fails to end in a subroutine name found in the `terminal_funcs` list.  So before the PR the critic would raise a criticism.

After the PR, and with the config `terminal_meths = logconfess` any method named `logconfess` counts as terminal.
So now the code is valid, so no criticism raised.

@petdance I missed that there were existing unit tests, but now found them.  I'll try to put something together. As you say positive and negative.

**Comment by xenomorpheus** [comment]
@petdance @robinsmidsrod I've made some changes and the tests are currently running (they pass locally). Doco has also been added.

Do I need to somehow redo this entire pull request so that it is only one commit ?

Please let me know the process if the current form isn't acceptable.

https://github.com/Perl-Critic/Perl-Critic/pull/935

**Comment by robinsmidsrod** [comment]
@xenomorpheus I think you still haven't fixed the rename from `terminal_meths` to `terminal_methods`, as mentioned by @petdance earlier. Other than that it looks okay to my eyes (although I haven't tested it).

**Comment by xenomorpheus** [comment]
> @xenomorpheus I think you still haven't fixed the rename from `terminal_meths` to `terminal_methods`, as mentioned by @petdance earlier. Other than that it looks okay to my eyes (although I haven't tested it).

Hi @robinsmidsrod I've addressed that in my earlier message and I am waiting to hear back from @petdance if he wishes to follow the established naming style or introduce minor change in style.

So I believe I've done all I can until I hear back from @petdance

**Comment by petdance** [comment]
I am getting updates on this, and have not forgotten it.  It's on my to-do list.

**Comment by xenomorpheus** [comment]
Hi @petdance I hope to hear from you when you are able.

**Comment by xenomorpheus** [comment]
Hi @petdance I hope to hear from you when you are able.

**Comment by xenomorpheus** [comment]
Hi @petdance I hope to hear from you when you are able.

**Comment by petdance** [comment]
> I hope to hear from you when you are able.

Yes, indeed.  That is when you will hear from me.  This issue is tops on my Perl-Critic to-do list.

**Comment by petdance** [comment]
This has been added at 1485fcc3c91228640b476bd9990455ed9c439f48.  It will go out in the next dev release, 1.139_02.  Thank you.

**Comment by petdance** [comment]
This is now up on CPAN in 1.140.

**Comment by xenomorpheus** [comment]
Thank you for your work. Very appreciated.  I apologise I haven't responded to your very kind offer for a commit. That is super cool of you.  I'll hope to get a commit sorted out one day. Again thank you for your hard work.
